src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -310,22 +316,19 @@ matchJoin
        , Show (variable Object)
        , Show (variable Meta)
        , SortedVariable variable
+       , MonadCounter m
        )
     => MetadataTools level StepperAttributes
     -> Map.Map (variable level) (variable level)
     -> [(PureMLPattern level variable, PureMLPattern level variable)]
-    -> Maybe
-        (Simplifier (PredicateSubstitution level variable))
-matchJoin tools quantifiedVariables patterns = do -- Maybe monad
+    -> MaybeT m (PredicateSubstitution level variable)
+matchJoin tools quantifiedVariables patterns = do
     matchedCounters <-
```

Comment:
With the current implementation, the `Counters` suffix does not make sense. Either `matched` or `matchedPatterns`